src/aleph/vm/controllers/qemu/instance.py

```diff
@@ -0,0 +1,348 @@
+                "create",
+                "-f",  # Format
+                "qcow2",
+                "-F",
+                parent_format,
+                "-b",
+                str(parent_image_path),
+                str(dest_path),
+            ]
+        )
+        return dest_path
+
+
+ConfigurationType = TypeVar("ConfigurationType")
+
+
+class AlephQemuInstance(Generic[ConfigurationType], CloudInitMixin, AlephControllerInterface):
+    vm_id: int
+    vm_hash: ItemHash
+    resources: AlephQemuResources
+    enable_console: bool
+    enable_networking: bool
+    hardware_resources: MachineResources
+    tap_interface: Optional[TapInterface] = None
+    vm_configuration: Optional[ConfigurationType]
+    is_instance: bool
+    qemu_process: Optional[Process]
+    support_snapshot = False
+    qmp_socket_path = None
+
+    def __repr__(self):
+        return f"<AlephQemuInstance {self.vm_id}>"
+
+    def __str__(self):
+        return f"vm-{self.vm_id}"
+
+    def __init__(
+        self,
+        vm_id: int,
+        vm_hash: ItemHash,
+        resources: AlephQemuResources,
+        enable_networking: bool = False,
+        enable_console: Optional[bool] = None,
+        hardware_resources: MachineResources = MachineResources(),
+        tap_interface: Optional[TapInterface] = None,
+    ):
+        self.vm_id = vm_id
+        self.vm_hash = vm_hash
+        self.resources = resources
+        if enable_console is None:
+            enable_console = settings.PRINT_SYSTEM_LOGS
+        self.enable_console = enable_console
+        self.enable_networking = enable_networking and settings.ALLOW_VM_NETWORKING
+        self.hardware_resources = hardware_resources
+        self.tap_interface = tap_interface
+
+    def to_dict(self):
+        """Dict representation of the virtual machine. Used to record resource usage and for JSON serialization."""
+        if self.qemu_process and psutil:
+            # The firecracker process is still running and process information can be obtained from `psutil`.
+            try:
+                p = psutil.Process(self.qemu_process.pid)
+                pid_info = {
+                    "status": p.status(),
+                    "create_time": p.create_time(),
+                    "cpu_times": p.cpu_times(),
+                    "cpu_percent": p.cpu_percent(),
+                    "memory_info": p.memory_info(),
+                    "io_counters": p.io_counters(),
+                    "open_files": p.open_files(),
+                    "connections": p.connections(),
+                    "num_threads": p.num_threads(),
+                    "num_ctx_switches": p.num_ctx_switches(),
+                }
+            except psutil.NoSuchProcess:
+                logger.warning("Cannot read process metrics (process %s not found)", self.qemu_process)
+                pid_info = None
+        else:
+            pid_info = None
+
+        return {
+            "process": pid_info,
+            **self.__dict__,
```

Comment:
Having all the `__dict__` values unpack into this can be dangerous. Defining `process` anywhere as a field can completely override your collected info. Maybe consider not unpacking it and instead make another field `properties` or similar

Reply:
One thing that bother me is that we endup in the `/about/executions` endpoint` with stuff like in the
```tap_interface	"<aleph.vm.network.interfaces.TapInterface object at 0x7fb95d0cffa0>"``` 

But it's totally outside of the scope of this pr

Reply:
https://github.com/aleph-im/aleph-vm/issues/494